### DISCUSSION TITLE
LIBUSB_ERROR_TIMEOUT where libusb seems to get stuck for roughly 168 seconds

### DISCUSSION THREAD

**Discussion by azlamzn** [libusb/libusb#1393] [Q&A]
OS: 
```
Ubuntu 20.04.4 LTS
```

Libraries:
```
libusb-1.0-0/focal,now 2:1.0.23-2build1 amd64 [installed,automatic]
libccid/focal,now 1.4.31-1 amd64 [installed,automatic]
libpcsclite1/focal,now 1.8.26-3 amd64 [installed,automatic]
```

Background:
I have a Python application using `pyscard` to read data. There is a seemingly random issue where libusb gets stuck during a transmit call for about 168 seconds which results in a LIBUSB_ERROR_TIMEOUT.

I have attached libusb debug logs with converted timestamps in order to provide some background around the series of events [libusb_debug_logs_12_05_2023.txt](https://github.com/libusb/libusb/files/13642943/libusb_debug_logs_12_05_2023.txt).

The issue occurs near the bottom of the file. At timestamp `2023-12-05 09:45:56` the Python application starts the transmit that ultimately times out. The error I can observe is at `Dec-05 09:55:45`, which is a time period where the device is unresponsive. The logs end with unplugging the device and the issue is resolved once the device is replugged.

Question:
I wanted to confirm if this timeout behavior for 168 seconds is a known pattern and what the recommended recovery path should be since a usb reset from the Python code does not help. Thank you so much!

**Comment by tormodvolden** [comment]
I don't know where the 168 seconds come from but I see in the log:
`[add_to_flying_list	arm	timerfd	for	timeout	in	167723ms	(first	in	line)`
and these show up long before the issue occurs. What is the timeout that you specify, if any?

Would it be possible for you try 1.0.27-rc1 and also updated pyusb?

**Reply by beedell-roke** [reply]
@tormodvolden and @azlamzn, the worst slowdowns that I have experienced are (per `sudo strace -Ttr lsusb`):

1.	<blockquote>

	~~~MA
	13:05:24 (+     0.000028) poll([{fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=8, events=POLLOUT}], 3, 60000) = 1 ([{fd=5, revents=POLLIN}]) <5.000064>
	13:05:29 (+     5.000310) ioctl(8, USBDEVFS_DISCARDURB, 0x5634727d8dc0) = 0 <0.000150>
	~~~

	</blockquote>

1.	<blockquote>

	~~~MA
	13:05:29 (+     0.000038) poll([{fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=8, events=POLLOUT}], 3, 60000) = 1 ([{fd=5, revents=POLLIN}]) <5.001474>
	13:05:34 (+     5.001697) ioctl(8, USBDEVFS_DISCARDURB, 0x5634727d8dc0) = 0 <0.000241>
	~~~

	</blockquote>

1.	<blockquote>

	~~~MA
	13:05:34 (+     0.000042) poll([{fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=8, events=POLLOUT}], 3, 60000) = 1 ([{fd=5, revents=POLLIN}]) <4.999935>
	13:05:39 (+     5.000037) ioctl(8, USBDEVFS_DISCARDURB, 0x5634727d8dc0) = 0 <0.000096>
	~~~

	</blockquote>

1.	<blockquote>

	~~~MA
	13:05:39 (+     0.000041) poll([{fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=8, events=POLLOUT}], 3, 60000) = 1 ([{fd=5, revents=POLLIN}]) <5.000808>
	13:05:44 (+     5.001124) ioctl(8, USBDEVFS_DISCARDURB, 0x5634727d8dc0) = 0 <0.000305>
	~~~

	</blockquote>

1.	<blockquote>

	~~~MA
	13:05:44 (+     0.000021) close(8)      = 0 <30.804063>
	13:06:15 (+    30.804300) write(1, "\n", 1
	~~~

	</blockquote>

However, I also see `LIBUSB_ERROR_TIMEOUT`:

<blockquote>

~~~MA
13:05:34 (+     0.000224) timerfd_settime(5, 0, {it_interval={tv_sec=0, tv_nsec=0}, it_value={tv_sec=0, tv_nsec=0}}, NULL) = 0 <0.000021>
13:05:34 (+     0.000075) ioctl(8, USBDEVFS_REAPURBNDELAY, 0x7ffe07466310) = -1 EAGAIN (Resource temporarily unavailable) <0.000017>
13:05:34 (+     0.000065) write(2, "can't get hub descriptor, LIBUSB"..., 82can't get hub descriptor, LIBUSB_ERROR_TIMEOUT (Resource temporarily unavailable)
~~~

</blockquote>

I have reposted this at https://github.com/libusb/libusb/issues/1344#issuecomment-3314935845. [^1]

[^1]: [`discussions/1343#discussioncomment-7658800`][1]

[1]: https://github.com/libusb/libusb/discussions/1343#discussioncomment-7658800

#### My Environment

1.	~~~sh
	#!/usr/bin/env sh
	rpm -qi $(rpm -qf $(command -v lsusb))
	~~~

1.	<blockquote>

	~~~YAML
	Name        : usbutils
	Version     : 018
	Release     : 3.fc42
	Architecture: x86_64
	Install Date: Fri 18 Apr 2025 15:28:13 BST
	Size        : 325356
	Signature   : RSA/SHA256, Fri 14 Mar 2025 10:57:29 GMT, Key ID c8ac4916105ef944
	Source RPM  : usbutils-018-3.fc42.src.rpm
	Build Date  : Fri 14 Mar 2025 09:10:08 GMT
	Build Host  : buildhw-x86-11.iad2.fedoraproject.org
	Packager    : Fedora Project
	Vendor      : Fedora Project
	URL         : http://www.linux-usb.org/
	Bug URL     : https://bugz.fedoraproject.org/usbutils
	~~~

	</blockquote>

**Comment by azlamzn** [comment]
Thanks for the quick response! Which version of pyusb are you referring to here (I am using 1.2.1)?

**Comment by tormodvolden** [comment]
So you are not using the stock python3-usb package then. Well, you can try latest pyusb git master, but I don't think anything relevant changed since 1.2.1.

OTOH a lot changed in libusb since 1.0.23. So if we are going to debug this further we need to use latest libusb git or RC.

**Comment by robin75411** [comment]
Thanks @tormodvolden. Is the latest version for libusb backward compatible with python 3.8?

**Comment by mcuee** [comment]
> Thanks @tormodvolden. Is the latest version for libusb backward compatible with python 3.8?

Yes.
